### PR TITLE
Added registration and login backend functionality

### PR DESCRIPTION
**Pull request by tmclean15**
Several packages were pulled in to accomplish the functionality, including:
-flask_bcrypt (password hashing)
-flask_jwt_extended (jwt functionality)
Validation was added to the username and login_email fields in the user model for data integrity. Changes were made to some of the column constraints in the user model for additional data integrity and consistency.The server Dockerfile was changed to check for any changes to the db_models, generate new migration scripts, and update the database on container startup.

### CODE REVIEW

File: server/Dockerfile [hunk 1]
```diff
@@ -7,6 +7,8 @@ COPY . .
 RUN pip3 install pipenv
 RUN pipenv install
 
-#CMD ["pipenv", "run", "flask", "run", "--host", "0.0.0.0"]
 
-CMD pipenv run flask db upgrade && pipenv run flask run --host=0.0.0.0
\ No newline at end of file
+CMD pipenv run flask db stamp head && \
```
Comment (MalcolmAG) [inline server/Dockerfile:11]: Clarification: What's the purpose of "stamp head" here?

File: server/db_models/pluser.py [hunk 2]
```diff
@@ -1,13 +1,42 @@
...
+            raise AssertionError('No password provided')
+        if len(password) < 6 or len(password) > 64:
+            raise AssertionError('Password must be between 6 and 64 characters')
+
+        hash = bcrypt.generate_password_hash(password.encode('utf-8'))
+        self.password_hash = hash.decode('utf-8')
+
+    @validates('username')
+    def validate_username(self, key, username):
+        if len(username) < 3 or len(username) > 64:
```
Comment (MalcolmAG) [inline server/db_models/pluser.py:29]: Should this be "len(username) > 20" ?

**Review by bonnieli — COMMENTED** [review]
very nice!

File: server/db_models/pluser.py [hunk 3]
```diff
@@ -1,13 +1,42 @@
...
+    
+    def set_password(self, password):
+        if not password:
+            raise AssertionError('No password provided')
+        if len(password) < 6 or len(password) > 64:
+            raise AssertionError('Password must be between 6 and 64 characters')
+
+        hash = bcrypt.generate_password_hash(password.encode('utf-8'))
+        self.password_hash = hash.decode('utf-8')
+
```
Comment (MalcolmAG) [inline server/db_models/pluser.py:26]: A small issue I have is that when you register, the register_handler does a validation, and the db does a validation. I'm wondering what's the purpose of having two?

**Review by MalcolmAG — APPROVED** [review]

File: server/db_models/pluser.py (hunk 2 above)
Comment (tmclean15) [reply server/db_models/pluser.py:29]: The username column currently has a length of 64 but I'm open to lowering it.

File: server/db_models/pluser.py (hunk 3 above)
Comment (tmclean15) [reply server/db_models/pluser.py:26]: The main purpose was to ensure data integrity for any future uses of the user model outside of the register_handler. On that note, we should probably add some validation for other fields in a future PR as the model validations I added were not comprehensive.

File: server/db_models/pluser.py (hunk 2 above)
Comment (MalcolmAG) [reply server/db_models/pluser.py:29]: It doesn't matter. 64 works

File: server/Dockerfile (hunk 1 above)
Comment (tmclean15) [reply server/Dockerfile:11]: Here is a good discussion of the issue.  https://stackoverflow.com/questions/17768940/target-database-is-not-up-to-date

In short, it indicates that the current state of the database represents the application of all migrations.

**Pull request merged**